docs/concepts/core-workflow-concepts.md

```diff
@@ -0,0 +1,129 @@
+# Core Concepts
+
+## Introduction
+
+This document explains the core concepts that are needed to understand the workflow engine.
+
+## Workflow
+
+A workflow is the definition of a process. It is made up of states and transitions between those states. The following diagram shows a simple workflow representing the lifetime of an item in a catalogue:
+
+```
+                  Create
+                     |
+                     | 
+                     |
+             +-------v-------+
+             |               |
+             | Waiting for   |
+      +------+ documentation |
+      |      |               |
+      |      +-------+-------+
+      |              |
+Edit                 | Publish
+(incomplete)         |
+      |      +-------v-------+               +---------------+
+      |      |               |               |               |
+      +------+   Published   |  Deprecate    |  Deprecated   |
+             |               +--------------->    (end)      |
+      +------>               |               |               |
+      |      +--+---------+--+               +---------------+
+      |         |         |
+ Edit +---------+         | Delete
+ (complete)               |
+             +------------v--+
+             |               |
+             |    Deleted    |
+             |     (end)     |
+             |               |
+             +---------------+
+
+```
+
+This workflow contains four states and 5 transitions. The states are represented by boxes, and the transitions by the lines between them.
+
+As you can see from the diagram, the workflow starts with the "Waiting for documentation" state and finishes when either the "Deprecated" or "Deleted" states are reached. Transitions allow the item to cycle between the "Published" and "Waiting for documentation" states.
+
+## Workflow Instance
+
+A workflow instance is a representation of a specific item's progress through the workflow. The workflow above represents the lifetime of an item in a catalogue; each of those items will be associated with a workflow instance representing its state at any point.
+
+A workflow instance contains the id and version of its parent workflow, the current state of the item in the workflow, and any other useful data that's associated with that instance.
+
+Instances also have a Status. The main statuses are:
+- Initializing, which indicates that a workflow instance has been created but is not yet ready to receive triggers 
```

Comment:
When a new workflow instance is created, it's immediately persisted with it's status set to "Initializing". Then, it's initialized properly - state set, entry conditions for that state evaluated, and entry actions executed. Then it's updated in the store.

I remember we had to make this change, but I'm currently struggling to remember why.

Reply:
Ah, I have remembered. It was for a slightly unusual edge case when using out-of-process hosting where the entry actions for the workflow's initial state cause a trigger to be raised that's targetted at the new instance (this might happen if the data supplied with the request to start the new instance satisfied some condition that meant the instance should move into the next state immediately. The only way to achieve this state change is via a transition, so a trigger is sent.

If further initialisation processing is required, then while this is happening the new trigger will be processed by the message processing host. The processing host attempts to find workflow instances to pass the trigger to by looking in the instance store. If we hadn't already saved the workflow instance, then the message processing host would not be able to identify it as a potential target for the trigger, which would then be discarded.

Reply:
So this arises because we allow entry actions for the initial state, and we don't consider ourselves to be in the initial state until we've completed those entry actions?

(I find the idea of entry conditions for the initial state a bit weird by the way. What happens if those fail? Do we fault?)

And the solution is essentially to add another state, but one that's intrinsic to the way workflows execute rather than defined explicitly in individual workflows—this `Initializing` state is when we're in the act of transitioning into the initial state, but haven't yet finished executing entry actions.

What's our mechanism for dealing with triggers that come in when we've not finished dealing with the previous trigger? Presumably we don't want to be trying to process two triggers concurrently for the same instance—do we take a lease or something? Couldn't we use the same approach here? Any triggers that happen while we're in the process of creation should have to wait until that creation process finishes just like they would if they happened while some other trigger were being processed. This feels like a specific instance of a more general problem (a trigger happens when we hadn't yet settled into a steady state after the preceding stimulus), so I don't get why we need a special mechanism for this for the initial creation.

Reply:
It's not specifically about the state it's in.

The message processing part of the solution is responsible for identifying candidate workflow instances to pass incoming triggers to. It does this by querying the workflow instance store for instances where the subjects/interests align.

If the new instance doesn't exist in the store, the message processing host won't identify it as a candidate instance for the incoming trigger, so it will never get the chance to process that trigger.

In terms of concurrency, yes, we take a lease on workflow instances whilst they are processing a trigger. But that only happens inside the engine when an instance is about to process a trigger - and in the particular scenario we're talking about, we'd never actually get that far.